cache/refs.go

```diff
@@ -203,103 +330,42 @@ func (cr *cacheRecord) size(ctx context.Context) (int64, error) {
 							continue
 						}
 						if info, err := cr.cm.ContentStore.Info(ctx, cdgst); err == nil {
-							usage.Size += info.Size
+							totalSize += info.Size
 						}
 					}
 				}
 			}
 		}
 		cr.mu.Lock()
-		cr.queueSize(usage.Size)
+		cr.queueSize(totalSize)
 		if err := cr.commitMetadata(); err != nil {
 			cr.mu.Unlock()
 			return s, err
 		}
 		cr.mu.Unlock()
-		return usage.Size, nil
+		return totalSize, nil
 	})
 	if err != nil {
 		return 0, err
 	}
 	return s.(int64), nil
 }
 
-func (cr *cacheRecord) parentRef(hidden bool, descHandlers DescHandlers) *immutableRef {
-	p := cr.parent
-	if p == nil {
-		return nil
-	}
-	p.mu.Lock()
-	defer p.mu.Unlock()
-	return p.ref(hidden, descHandlers)
-}
-
-// must be called holding cacheRecord mu
-func (cr *cacheRecord) mount(ctx context.Context, readonly bool) (snapshot.Mountable, error) {
-	if cr.mutable {
-		m, err := cr.cm.Snapshotter.Mounts(ctx, cr.getSnapshotID())
-		if err != nil {
-			return nil, errors.Wrapf(err, "failed to mount %s", cr.ID())
-		}
-		if readonly {
-			m = setReadonly(m)
-		}
-		return m, nil
-	}
-
-	if cr.equalMutable != nil && readonly {
-		m, err := cr.cm.Snapshotter.Mounts(ctx, cr.equalMutable.getSnapshotID())
-		if err != nil {
-			return nil, errors.Wrapf(err, "failed to mount %s", cr.equalMutable.ID())
-		}
-		return setReadonly(m), nil
-	}
-
-	if err := cr.finalize(ctx); err != nil {
-		return nil, err
-	}
-	if cr.viewMount == nil { // TODO: handle this better
-		view := identity.NewID()
-		l, err := cr.cm.LeaseManager.Create(ctx, func(l *leases.Lease) error {
-			l.ID = view
-			l.Labels = map[string]string{
-				"containerd.io/gc.flat": time.Now().UTC().Format(time.RFC3339Nano),
-			}
-			return nil
-		}, leaseutil.MakeTemporary)
-		if err != nil {
-			return nil, err
-		}
-		ctx = leases.WithLease(ctx, l.ID)
-		m, err := cr.cm.Snapshotter.View(ctx, view, cr.getSnapshotID())
-		if err != nil {
-			cr.cm.LeaseManager.Delete(context.TODO(), leases.Lease{ID: l.ID})
-			return nil, errors.Wrapf(err, "failed to mount %s", cr.ID())
-		}
-		cr.view = view
-		cr.viewMount = m
-	}
-	return cr.viewMount, nil
-}
-
 // call when holding the manager lock
 func (cr *cacheRecord) remove(ctx context.Context, removeSnapshot bool) error {
 	delete(cr.cm.records, cr.ID())
-	if cr.parent != nil {
-		cr.parent.mu.Lock()
-		err := cr.parent.release(ctx)
-		cr.parent.mu.Unlock()
-		if err != nil {
-			return err
-		}
-	}
 	if removeSnapshot {
-		if err := cr.cm.LeaseManager.Delete(context.TODO(), leases.Lease{ID: cr.ID()}); err != nil {
-			return errors.Wrapf(err, "failed to remove %s", cr.ID())
+		if err := cr.cm.LeaseManager.Delete(ctx, leases.Lease{
+			ID: cr.ID(),
+		}); err != nil {
+			return errors.Wrapf(err, "failed to delete lease for %s", cr.ID())
 		}
 	}
 	if err := cr.cm.MetadataStore.Clear(cr.ID()); err != nil {
-		return err
+		return errors.Wrapf(err, "failed to delete metadata of %s", cr.ID())
+	}
+	if err := cr.parentRefs.release(ctx); err != nil {
```

Comment:
My thinking was that if there was a disk error on deletion it could potentially leak some data for the current snapshot but at least not mess up the reference count for the whole parent chain. I guess neither is perfect. I think this code used to call `snapshot.Delete` directly, making errors more likely.